### PR TITLE
SIP-341: Add Configurer Address (#2061)

### PR DESCRIPTION
**Pull request by noahlitvin**
* saving work so far

* contracts draft finished

* added set configurer and changed fn name

* changed to configurable, context msg sender, and seperated contracts

* removed setConfigurer from ownable

* before a configurer is set

* added edge case protection

* tests

* added edge case test

* added tests

* added comment

* 100% coverage

* removed modifier onlyOwnerOrConfigurer

* Update utils/core-contracts/test/contracts/ownership/Configurable.test.ts

* audit refactor

---------

### CODE REVIEW

File: utils/core-contracts/contracts/ownership/Configurable.sol [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    function acceptConfigurerRole() public override {
+        ConfigurableStorage.ConfigurerData storage store = ConfigurableStorage.loadConfigurer();
+
+        address currentNominatedConfigurer = store.nominatedConfigurer;
+        if (ERC2771Context._msgSender() != currentNominatedConfigurer) {
+            revert NotNominatedAsConfigurer(ERC2771Context._msgSender());
+        }
+
+        emit ConfigurerChanged(store.configurer, currentNominatedConfigurer);
+        store.configurer = currentNominatedConfigurer;
```
Comment (noisekit) [inline utils/core-contracts/contracts/ownership/Configurable.sol:29]: Shouldn't event be emitted after change happend?
```suggestion
        store.configurer = currentNominatedConfigurer;
        emit ConfigurerChanged(store.configurer, currentNominatedConfigurer);
```

File: utils/core-contracts/contracts/ownership/Configurable.sol [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            revert ChangeError.NoChange();
+        }
+
+        if (store.nominatedConfigurer != address(0)) {
+            store.nominatedConfigurer = address(0); // reset nominated configurer to address(0)
+            emit ConfigurerNominated(address(0));
+        }
+
+        emit ConfigurerChanged(store.configurer, newConfigurer);
+        store.configurer = newConfigurer;
```
Comment (noisekit) [inline utils/core-contracts/contracts/ownership/Configurable.sol:85]:
```suggestion
        store.configurer = newConfigurer;
        emit ConfigurerChanged(store.configurer, newConfigurer);
```

File: utils/core-contracts/contracts/ownership/ConfigurableStorage.sol [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import "./OwnableStorage.sol";
+
+library ConfigurableStorage {
+    bytes32 private constant _SLOT_CONFIGURER_STORAGE =
+        keccak256(abi.encode("io.synthetix.core-contracts.Configurer"));
+
+    struct ConfigurerData {
+        address configurer;
+        address nominatedConfigurer;
+    }
```
Comment (noisekit) [inline utils/core-contracts/contracts/ownership/ConfigurableStorage.sol:15]: Is there any particular reason to have a struct and not 2 separate properties?

```suggestion
    address configurer;
    address nominatedConfigurer;
```

**Review by noisekit — APPROVED** [review]
A couple of minor questions, but overall looks good to me.
